app/src/main/java/io/homeassistant/companion/android/util/TLSWebViewClient.kt

```diff
@@ -0,0 +1,106 @@
+import android.webkit.ClientCertRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.annotation.RequiresApi
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.lifecycleScope
+import io.homeassistant.companion.android.common.data.keychain.KeyChainRepository
+import kotlinx.coroutines.launch
+import java.security.Principal
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+
+open class TLSWebViewClient @Inject constructor(private var keyChainRepository: KeyChainRepository) : WebViewClient() {
+
+    var isTLSClientAuthNeeded = false
+        private set
+
+    var hasUserDeniedAccess = false
+        private set
+
+    private var key: PrivateKey? = null
+    private var chain: Array<X509Certificate>? = null
+
+    private fun getActivity(context: Context?): Activity? {
+        if (context == null) {
+            return null
+        } else if (context is ContextWrapper) {
+            return if (context is Activity) {
+                context
+            } else {
+                getActivity(context.baseContext)
+            }
+        }
+        return null
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        // Let the WebViewActivity know the endpoint requires TLS Client Auth
+        isTLSClientAuthNeeded = true
+
+        // Aim to obtain the private key for the whole lifecycle of the WebViewActivity
+        val activity = getActivity(view.context)
+        if (activity == null || activity !is AppCompatActivity) return
+
+        activity.lifecycleScope.launch {
+            // If the key is available, process the request
+            if (key != null && chain != null) {
+                request.proceed(key, chain)
+            } else {
+                // Get the key and the chain from the repo (if the user previously chose)
+                key = keyChainRepository.getPrivateKey()
+                chain = keyChainRepository.getCertificateChain()
+
+                if (key != null && chain != null) {
+                    request.proceed(key, chain)
+                } else {
+                    // If no key is available, then the user must be prompt for a key
+                    // The whole operation is wrapped in the selectPrivateKey method but caution as it must occurs outside of the main thread
+                    // see: https://developer.android.com/reference/android/security/KeyChain#getPrivateKey(android.content.Context,%20java.lang.String)
+                    selectClientCert(activity, request.principals, request)
+                }
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun selectClientCert(activity: Activity, principals: Array<Principal>?, request: ClientCertRequest) {
+        if (activity !is AppCompatActivity) {
+            return
+        }
```

Comment:
I think we could be more strict, because it _should_ be an `AppCompatActivity`, right? What do you think?

But I'm also curious about @JBassett's opinion about this.

```suggestion
        require(activity is AppCompatActivity)
```

Reply:
Looks fine be, done.